rollup/pipeline/pipeline.go

```diff
@@ -0,0 +1,408 @@
+	ErrApplyStageDone           = errors.New("apply stage is done")
+	ErrUnexpectedL1MessageIndex = errors.New("unexpected L1 message index")
+
+	lifetimeTimer   = metrics.NewRegisteredTimer("pipeline/lifetime", nil)
+	applyTimer      = metrics.NewRegisteredTimer("pipeline/apply", nil)
+	applyIdleTimer  = metrics.NewRegisteredTimer("pipeline/apply_idle", nil)
+	applyStallTimer = metrics.NewRegisteredTimer("pipeline/apply_stall", nil)
+	cccTimer        = metrics.NewRegisteredTimer("pipeline/ccc", nil)
+	cccIdleTimer    = metrics.NewRegisteredTimer("pipeline/ccc_idle", nil)
+)
+
+type Pipeline struct {
+	chain    *core.BlockChain
+	vmConfig *vm.Config
+	parent   *types.Block
+	start    time.Time
+
+	// accumalators
+	ccc            *circuitcapacitychecker.CircuitCapacityChecker
+	Header         types.Header
+	state          *state.StateDB
+	nextL1MsgIndex uint64
+	blockSize      common.StorageSize
+	txs            types.Transactions
+	coalescedLogs  []*types.Log
+	receipts       types.Receipts
+	gasPool        *core.GasPool
+
+	// com channels
+	TxnQueue         chan *types.Transaction
+	applyStageRespCh <-chan error
+	ResultCh         <-chan *Result
+
+	// Test hooks
+	beforeTxHook func() // Method to call before processing a transaction.
+}
+
+func NewPipeline(
+	chain *core.BlockChain,
+	vmConfig *vm.Config,
+	state *state.StateDB,
+
+	header *types.Header,
+	ccc *circuitcapacitychecker.CircuitCapacityChecker,
+) *Pipeline {
+	parent := chain.GetBlock(header.ParentHash, header.Number.Uint64()-1)
+	return &Pipeline{
+		chain:          chain,
+		vmConfig:       vmConfig,
+		parent:         parent,
+		nextL1MsgIndex: *rawdb.ReadFirstQueueIndexNotInL2Block(chain.Database(), parent.Hash()),
+		Header:         *header,
+		ccc:            ccc,
+		state:          state,
+		gasPool:        new(core.GasPool).AddGas(header.GasLimit),
+	}
+}
+
+func (p *Pipeline) WithBeforeTxHook(beforeTxHook func()) *Pipeline {
+	p.beforeTxHook = beforeTxHook
+	return p
+}
+
+func (p *Pipeline) Start(deadline time.Time) error {
+	p.start = time.Now()
+	p.TxnQueue = make(chan *types.Transaction)
+	applyStageRespCh, candidateCh, err := p.traceAndApplyStage(p.TxnQueue)
+	if err != nil {
+		log.Error("Failed starting traceAndApplyStage", "err", err)
+		return err
+	}
+	p.applyStageRespCh = applyStageRespCh
+	p.ResultCh = p.cccStage(candidateCh, deadline)
+	return nil
+}
+
+func (p *Pipeline) TryPushTxns(txs types.OrderedTransactionSet, onFailingTxn func(txnIndex int, tx *types.Transaction, err error) bool) *Result {
+	for {
+		tx := txs.Peek()
+		if tx == nil {
+			break
+		}
+
+		result, err := p.TryPushTxn(tx)
+		if result != nil {
+			return result
+		}
+
+		switch {
+		case err == nil, errors.Is(err, core.ErrNonceTooLow):
+			txs.Shift()
+		default:
+			if errors.Is(err, ErrApplyStageDone) || onFailingTxn(p.txs.Len(), tx, err) {
+				close(p.TxnQueue)
+				p.TxnQueue = nil
+				return nil
+			}
+
+			if tx.IsL1MessageTx() {
+				txs.Shift()
+			} else {
+				txs.Pop()
+			}
```

Comment:
Yeah, I went thru all of them. `txs.Shift()` is only relevant for L1 messages and ErrNonceTooLow errors.  

> .g. in the "strange error" branch in the original code, we'd txs.Shift() for both types

for example, in this case we will end up needing to Shift() all pending txns of an account anyways. Because once a txn fails, the rest will too because of their nonce.